plugins/newrelic/src/api/index.ts

```diff
@@ -71,39 +76,105 @@ export interface NewRelicApi {
+  relation: string;
+}
+
+interface NewRelicPageReadResult {
+  hasMoreApplications: boolean;
+  pagination: PaginationInformation | undefined;
+  applicationsFromReadPage: NewRelicApplication[];
+}
+
 export class NewRelicClient implements NewRelicApi {
   private readonly discoveryApi: DiscoveryApi;
+  private readonly fetchApi: FetchApi;
   private readonly proxyPathBase: string;
+  private baseUrl: string;
 
   constructor(options: Options) {
     this.discoveryApi = options.discoveryApi;
+    this.fetchApi = options.fetchApi;
     this.proxyPathBase = options.proxyPathBase ?? DEFAULT_PROXY_PATH_BASE;
+    this.baseUrl = '';
   }
 
   async getApplications(): Promise<NewRelicApplications> {
-    const url = await this.getApiUrl('apm', 'applications.json');
-    const response = await fetch(url);
-    let responseJson;
+    if (!this.baseUrl) {
+      const proxyUrl = await this.discoveryApi.getBaseUrl('proxy');
+      this.baseUrl = `${proxyUrl}${this.proxyPathBase}/apm/api/applications.json`;
+    }
+
+    let targetUrl = this.baseUrl;
+    let hasNextPage = true;
 
     try {
-      responseJson = await response.json();
+      let applications: NewRelicApplication[] = [];
+
+      do {
+        const { hasMoreApplications, pagination, applicationsFromReadPage } =
+          await this.fetchNewRelic(targetUrl);
+
+        hasNextPage = hasMoreApplications;
+        targetUrl = hasNextPage ? pagination!.nextPageLink : '';
+        applications = applications.concat(applicationsFromReadPage);
+      } while (hasNextPage);
+
+      return { applications };
     } catch (e) {
-      responseJson = { applications: [] };
+      return { applications: [] };
     }
+  }
+
+  private async fetchNewRelic(
+    targetUrl: string,
+  ): Promise<NewRelicPageReadResult> {
+    const response = await this.fetchApi.fetch(targetUrl);
+    const responseJson = await response.json();
 
-    if (response.status !== 200) {
+    if (!response.ok) {
       throw new Error(
         `Error communicating with New Relic: ${
           responseJson?.error?.title || response.statusText
         }`,
       );
     }
 
-    return responseJson;
+    const readResponse = responseJson as NewRelicApplications;
+    const linkHeader =
+      response.headers.get('Link') || response.headers.get('link');
+    const pagination = this.parseLinkHeader(linkHeader);
+
+    return {
+      hasMoreApplications: !!(pagination && pagination.relation === 'next'),
+      pagination,
+      applicationsFromReadPage: readResponse.applications,
+    };
   }
 
-  private async getApiUrl(product: string, path: string) {
-    const proxyUrl = await this.discoveryApi.getBaseUrl('proxy');
-    return `${proxyUrl}${this.proxyPathBase}/${product}/api/${path}`;
+  private parseLinkHeader(
+    linkHeader: string | null,
+  ): PaginationInformation | undefined {
+    if (!linkHeader) {
+      return undefined;
+    }
+
+    const nextRelevantLink = linkHeader.replaceAll(' ', '').split(',')[0];
```

Comment:
Small nit: there could actually be multiple separate Link headers, each with their different `rel` etc. You could do `for ( const [key, value] of response.headers.entries())` and iterate that, looking for keys equal to `link` (it's lowercased there). I see your split by a comma here compensates for the fact that the single `get` extracts a comma separated form but you could do them one by one looking for the right one instead of just taking the first

Reply:
Is the exact implementation of how it processes the headers browser defined? The interface makes it sound like it is a map and two separate keys with the same name should not be possible. 

When testing in Chrome, I see that it flattens duplicate header names into a comma separated list. In this example, the ```Testheader``` is duplicated:

![image](https://github.com/backstage/backstage/assets/129326788/f957162a-5269-4f8b-a0df-0f52359f3d92)

![image](https://github.com/backstage/backstage/assets/129326788/d1feae9c-f9c8-4efc-98bb-830ee69def23)

Reply:
Yep that's what I meant; the map accessor is a "simplified" way of accessing things and it glosses over the multiple-headers situation by string concatenation. Which sort of usually works, so it's not a huge deal, as long as you re-split it. But note that `headers.entries()` does NOT go via a map, it retains all of the original headers one by one, whether there were duplicates or not. Arguably cleaner, but more work to access them that way

Reply:
I am sorry, but I am still confused about this. I am seeing that ```headers.entries()``` also combines duplicate header names into a single comma separated string, like ```get``` does. The keys returned from ```headers.entries()``` look to be distinct, regardless of if there were duplicates. If we switch to ```headers.entries()```, wouldn't we still have to split it?

In the fetch specification, it sounds like [headers except for ```Set-Cookie``` are supposed to be combined when exposed to JavaScript](https://fetch.spec.whatwg.org/#terminology-headers). Is this something that not all browsers respect, or is ```.entries()``` supposed to be an exception?
>"A header list is essentially a specialized multimap: an ordered list of key-value pairs with potentially duplicate keys. Since headers other than `Set-Cookie` are always combined when exposed to client-side JavaScript, implementations could choose a more efficient representation, as long as they also support an associated data structure for `Set-Cookie` headers."

![image](https://github.com/backstage/backstage/assets/129326788/288727ad-9899-4326-a4ee-ef8d312e2059)
![image](https://github.com/backstage/backstage/assets/129326788/5212bad5-abea-4acb-a707-998f90374094)

Reply:
Oh alright! Granted, I only tested with the array-shaped Headers constructor and that one did return the individual items again IIRC. Well either way, I say I've been frustrating enough with the pokery here :) Thanks for entertaining the thought and being thorough. Go ahead and keep things as-is.

Reply:
That's alright 👍. I was just confused because ```headers.entries()``` wasn't behaving differently from ```get```. I just wanted to make sure I wasn't doing something wrong.